### DISCUSSION TITLE
v11: Updates and Breaking Changes

### DISCUSSION THREAD

**Discussion by chrtze** [xyflow/xyflow#2307] [New Features]
We are working on the next major release (v11) which will include some breaking changes. You can find all the features that we are planning in [our roadmap](https://github.com/orgs/wbkd/projects/1/views/7) and the current status in the [pull request](https://github.com/wbkd/react-flow/pull/2279).

The most noticeable change will be the new repo structure and build process: We want to split React Flow into individual packages, that can be installed separately. The npm package names will change from `react-flow-renderer` to `@react-flow/...`. The core component will stay mostly the same, but we want to move the additional components like minimap, background, controls to individual packages to iteratively reduce the size of the core.

The goal of this is to have a monorepo with individual packages instead of a single entrypoint which will make it easier for us to create plugins or additional components that can be consumed only if you need them. 

Another thing that we want to change is the handling of the default styles. We have opened a [separate discussion](#2306) for this topic.

There are some smaller fixes currently on the roadmap for v11 but we want to keep the API mostly the same, so that you can easily upgrade by only installing the new package.

We would be happy to hear your feedback and additional ideas on the changes we are planning for v11. Also, we will share progress in this discussion.

**Comment by moklick** [comment]
# React Flow v11 Beta

As explained above the major changes are under the hood and you won't notice them. ReactFlow is now a monorepo and split into packages, that you can install separately.

For now you can help us to test the next major release (v11) of ReactFlow by installing it via:

```sh 
npm install reactflow@next
```

> **Warning**
> ReactFlow v11 needs at least React v17.0.0


## Usage 

v11 doesn't inject styles automatically anymore. You need to load the styles on your own. For example like this:

```js
import ReactFlow, { Background, Controls, MiniMap } from 'reactflow';
import 'reactflow/dist/style.css';
```

## Breaking Changes: 

* changed package name from "react-flow-renderer" to "reactflow"
* you need to load the styles by yourself. Default theming : `'reactflow/dist/style.css'` or just the base styles: `'reactflow/dist/base.css'`
* React Flow props: merge `defaultPosition` and `defaultZoom` into `defaultViewport` #2313 
* React Flow props: remove `onConnectStop` and `onClickConnectStop` #2299
* ConnectionLine props: remove `sourceX`, `sourceY` .. use `fromX`, `fromY` and so on #2300
* Background props: rename `strokeWidth` to `lineWidth` #2320
* Removal of `getBezierEdgeCenter`, `getSimpleBezierEdgeCenter` and `getEdgeCenter`
    * From now only we not only returning the path, but also the center / label position from `getBezierPath`, `getSmoothStepPath` and `getStraightPath`

## New APIs and updates:

* Better accessibility: Nodes and edges are focusable, selectable, moveable and deleteable with the keyboard. `aria-` attributes for all elements
* React Flow props: `disableKeyboardA11y` - v11 comes with keyboard controls for selecting and moving nodes and edges. These can be removed with this prop #1033 
* React Flow props: `onPaneMouseEnter`, `onPaneMouseMove` and `onPaneMouseLeave` #2312
* React Flow props: `nodeOrigin` (default: `[0, 0]`) - prop to control the node origin, `[0.5, 0.5]` would be center, `[1, 1]` would be bottom right 
* Background props: new `BackgroundVariant.Cross` variant, `lineWidth` prop, `gap` can be `number` or `[number, number]` #2320
* Hooks: `useOnViewportChange` and `useOnSelectionChange`:
```js
useOnViewportChange({
    onStart: (viewport: Viewport) => console.log('start', viewport),
    onChange: (viewport: Viewport) => console.log('change', viewport),
    onEnd: (viewport: Viewport) => console.log('end', viewport)
});
```
* Hook: `useNodesInitialized` - returns true if all nodes are initialized and if there is more than one node
* Nodes and edges options: add `deletable` flag - `true` by default
* Edge option: `interactionWidth` (default = 20) - renders second invisible edge that makes it easier to interact with an edge
* Edge option: `pathOptions` can be passed for `smoothstep` (`{  offset: number, borderRadius: number }`) and `default`  (`{  curvature: number })`) edges
* Cursor = grabbing, while dragging a node or panning
* Better routing for smoothstep and step edges https://twitter.com/reactflowdev/status/1567535405284614145 
* Nicer edge updating behaviour https://twitter.com/reactflowdev/status/1564966917517021184
* Pane pannable with middle mouse button
* You can pan over nodes when they are not draggable (`draggable=false` or `nodesDraggable` false) 
* `<BaseEdge />` component gets exported that can be used to build custom edges
* Packages can be installed separately via `npm install @reactflow/core @reactflow/background @reactflow/controls @reactflow/minimap`

## Feedback

Let us know what you think! It's still in the making and we are open for changes :)

**Reply by cryptoHaHa** [reply]
Thank you for your big update!
I found one problem, but not sure if it's an issue or not.
In version 10, the below code snippet works:

```
const spacePressed = useKeyPress('Space');

<ReactFlow
  nodes={initialNodes}
  edges={initialEdges}
  nodeTypes={nodeTypes}
  fitView
  preventScrolling={false}
  zoomOnScroll={spacePressed}
  snapToGrid={true}
>
  <Background />
</ReactFlow>
```
I am testing version 11 and the above code doesn't work anymore.
To be exactly, when I disable the preventScrolling option, Space key + scroll zoom feature doesn't work.
I added reactflow element in the middle of page content, so I need to disable a preventScroll option.
I hope you will consider this problem.

Thank you!

**Reply by moklick** [reply]
Thanks for your feedback. I will check that.

**Reply by moklick** [reply]
This is a bug with v10. You need to set `preventScrolling` accordingly:

```js
const spacePressed = useKeyPress('Space');

<ReactFlow
  nodes={initialNodes}
  edges={initialEdges}
  nodeTypes={nodeTypes}
  fitView
  preventScrolling={spacePressed}
  zoomOnScroll={spacePressed}
  snapToGrid={true}
>
  <Background />
</ReactFlow>
```

Does this work for you?

**Comment by cryptoHaHa** [comment]
Yes, it works on my end correctly.
Thank you!.

**Comment by cryptoHaHa** [comment]
One more thing.
In version 11, centerY and centerX options don't work well in getSmoothStepPath function.
I made custom edge to change the smoothstep edge center positions. But if I add center positions, edge is changed strangely.
Please check that.
And I am not sure how use these options.
```
Edge option: interactionWidth (default = 20) - renders second invisible edge that makes it easier to interact with an edge
Edge option: pathOptions can be passed for smoothstep ({ offset: number, borderRadius: number }) and default ({ curvature: number })) edges
```
Where and how use these options?

Thank you again.
